cli/cli-extensions.md

```diff
@@ -0,0 +1,195 @@
+
+Built-in commands cannot be overridden by extensions, but subcommands can be added to existing commands.
+
+### Discovery and Help Text
+
+Running `temporal --help` (or `temporal help`) _does not_ list extensions. But running `temporal help -a` (or `--all`)
+_does_ list all extensions by traversing the `PATH` and getting every `temporal-`-prefixed executable. There is no short
+description for any of these extensions, they are simply shown as available external commands. All extensions on the
+`PATH` are shown as they are, even if they are multiple commands deep. For example `temporal-foo-bar` and
+`temporal-workflow-dosomething-somethingelse` are shown as `foo bar` and `workflow dosomething somethingelse` in the
+list respectively, not as just `foo` or implied as part of the `workflow` command.
+
+Extensions on built-in commands are not shown as part of the parent command's help.
+
+Running `temporal help <command sequence>` is treated as `temporal <command sequence> --help`. So the same style of
+lookup is performed for extensions.
+
+### Invocation and Flags
+
+Temporal CLI supports arbitrarily ordered flags. So `temporal --address foo workflow start`,
+`temporal workflow --address foo start`, and `temporal workflow start --address foo` are all the same. It is still a
+subject of active research on whether flags of _parent_ commands can still be placed before the extension command in the
+CLI. Regardless, flags of the extension command _must_ come after the extension command because CLI does not know
+whether a flag takes an argument or not so it cannot disambiguate. For example, CLI
+`temporal --myflag someval1 someval2` cannot determine whether that is `temporal someval1 someval2 --myflag` or
+`temporal someval2 --myflag someval1`. So the latter forms must be the forms used for extension-specific flags.
+
+All flags of the parent command _should_ be handled properly by the extension. This means even root-level flags. So for
+example, `temporal-foo` _should_ handle root-level flags like `--output`. However, this isn't a _must_ requirement. See
+the helper library section for some helpers.
+
+Currently, the only root-level flag that the parent `temporal` process respects when calling an extension is
+`--command-timeout` (even though it is still passed along). All other root-level flags are up to the extension to
+handle.
+
+Invocation of the extension is done as a subprocess. Stdin, stdout, stderr, exit code, etc are all handled by the
+extension and just relayed through the `temporal` process as is. It is an area of active research on whether interrupt
+signals can be ignored by the parent `temporal` process to be handled by the subprocess, though that is the goal.
+
+### Helper Library
+
+Built-in commands leverage several helpers to be consistent with the CLI ecosystem. Extension commands should be able to
+do the same, within reason, so a helper library will be made available. Originally it was thought that such a helper
```

Comment:
Can you outline how users who don't want to use Go will proceed? I expect Python will be a common language to want to implement extensions in.

Reply:
Like kubectl (https://kubernetes.io/docs/tasks/extend-kubectl/kubectl-plugins/#using-the-command-line-runtime-package), they will not get a helper library at this time. I can clarify here if needed that there are no plans for any helper libraries for CLI extensions beyond the language the CLI is written in. I actually don't expect Python will be a common language to implement extensions in (but Python may be a common language to write general CLIs, just not as extensions to the Temporal CLI in Go I don't think anymore than it is for `kubectl`).

Reply:
Python's a bit too slow to be a good language to write general CLIs in, but in this context where you know you're making a network call that disadvantage doesn't apply. It's fine if Go is the only offering initially but I don't think plugin authors should have to care which language the CLI is written in. And since plugins are often written by people to "scratch an itch" (often for their own development environments), I think it will be common to choose a more convenient language than Go.

Reply:
You are more than welcome to write it in any language. If you care about convenience, you may choose bash. But if you want a helper library, it's only reasonable for us to maintain it in Go at this time (and for our use, the `temporal-cloud` CLI, we'll of course only use Go). Maybe we can revisit if the desire is large enough for Temporal CLI extensions in Python, though I suspect like similar Go CLIs there won't be much desire.

Reply:
Personally I am fine only providing helpers for Go at this time.